dango/bank/src/execute.rs

```diff
@@ -248,34 +285,104 @@ fn decrease_supply(
 
 fn increase_balance(
     storage: &mut dyn Storage,
-    address: &Addr,
-    denom: &Denom,
+    address: Addr,
+    denom: Denom,
     amount: Uint128,
-) -> StdResult<Option<Uint128>> {
-    BALANCES.may_modify(storage, (address, denom), |maybe_balance| {
-        let balance = maybe_balance.unwrap_or(Uint128::ZERO).checked_add(amount)?;
-        // Only write to storage if the balance is non-zero.
-        if balance.is_zero() {
-            Ok(None)
-        } else {
-            Ok(Some(balance))
-        }
-    })
+) -> StdResult<Option<BalanceChange>> {
+    BALANCES
+        .may_modify(storage, (&address, &denom), |maybe_balance| {
+            let balance = maybe_balance.unwrap_or(Uint128::ZERO).checked_add(amount)?;
+            // Only write to storage if the balance is non-zero.
+            if balance.is_zero() {
+                Ok(None)
+            } else {
+                Ok(Some(balance))
+            }
+        })
+        .map(|maybe_balance| {
+            maybe_balance.map(|_| BalanceChange {
+                address,
+                denom,
+                amount,
+                direction: BalanceChangeDirection::Increase,
+            })
+        })
 }
 
 fn decrease_balance(
     storage: &mut dyn Storage,
-    address: &Addr,
-    denom: &Denom,
+    address: Addr,
+    denom: Denom,
     amount: Uint128,
-) -> StdResult<Option<Uint128>> {
-    BALANCES.may_modify(storage, (address, denom), |maybe_balance| {
-        let balance = maybe_balance.unwrap_or(Uint128::ZERO).checked_sub(amount)?;
-        // If balance is reduced to zero, delete it, to save disk space.
-        if balance.is_zero() {
-            Ok(None)
-        } else {
-            Ok(Some(balance))
+) -> StdResult<Option<BalanceChange>> {
+    BALANCES
+        .may_modify(storage, (&address, &denom), |maybe_balance| {
+            let balance = maybe_balance.unwrap_or(Uint128::ZERO).checked_sub(amount)?;
+            // If balance is reduced to zero, delete it, to save disk space.
+            if balance.is_zero() {
+                Ok(None)
+            } else {
+                Ok(Some(balance))
+            }
+        })
+        .map(|maybe_balance_change| {
+            maybe_balance_change.map(|_| BalanceChange {
+                address,
+                denom,
+                amount,
+                direction: BalanceChangeDirection::Decrease,
+            })
+        })
+}
+
+#[derive(Default)]
+struct EventBuilder {
+    user_changes: BTreeMap<Addr, BTreeMap<Denom, (Uint128, BalanceChangeDirection)>>,
+}
+
+impl EventBuilder {
+    pub fn add_event(&mut self, event: Option<BalanceChange>) -> Result<(), StdError> {
+        if let Some(event) = event {
+            self.user_changes
+                .entry(event.address)
+                .or_default()
+                // There are no way to add multiple denom for the same address.
+                .insert(event.denom, (event.amount, event.direction));
```

Comment:
better to add a `debug_assert!` to ensure the return value of `insert` is `None`.